### DISCUSSION TITLE
How to migrate an inherited row

### DISCUSSION THREAD

**Discussion by Mar-he** [serenity-is/Serenity#6694] [Help]
I'm in the process of migrating a Serenity 3 application up to Serenity 6 and stumpled upon the following bit of code:

```cs
public class AuditLogRow : Row<AuditLogRow.RowFields>, IIdRow, INameRow
{
        [DisplayName("Id"), PrimaryKey, IdProperty]
        public Guid? Id
        {
            get => fields.Id[this];
            set => fields.Id[this] = value;
        }
       /* removed for brevity*/
       
        public AuditLogRow(RowFields fields)
            : base(fields)
        {
        }

       public AuditLogRow(RowFieldsBase fields)
          : base(fields)
        {
        }
}

public sealed class AnalysisOrderRow : AuditLogRow
{
        [IdProperty]
        [DisplayName("Id"), PrimaryKey]
        [Hidden]
        [Updatable(false)]
        public new Guid? Id
        {
            get => fields.Id[this];
            set => fields.Id[this] = value;
        }

        [DisplayName("PAZ"), Column("CAQReclamationNumber"), Size(255), NotNull, QuickSearch]
        [Updatable(false), NameProperty]
        public String CaqReclamationNumber
        {
            get => fields.CaqReclamationNumber[this];
            set => fields.CaqReclamationNumber[this] = value;
        }

        public new static readonly RowFields Fields = new RowFields().Init();

        public AnalysisOrderRow()
        {
        }

        public AnalysisOrderRow(RowFields fields)
            : base(fields)
        {
        }
}
```

AuditLogRow has an additional constructor (` public AuditLogRow(RowFieldsBase fields)`) which is used in AnalysisOrderRow. However, AnalysisOrderRow has the identical compiler error for each of its properties:

"AuditLogRow.RowFields" enthält keine Definition für "CaqReclamationNumber", und es konnte keine zugängliche CaqReclamationNumber-Erweiterungsmethode gefunden werden, die ein erstes Argument vom Typ "AuditLogRow.RowFields" akzeptiert (möglicherweise fehlt eine using-Direktive oder ein Assemblyverweis)"

Sorry, about the German, I can't change the machines system language. Basically it says, fields has no definition for CaqReclamationNumber.

Is it possible to inherit from an already existing row or should I ditch the inhertiance and move every property over?

**Comment by VictorTomaili** [comment]
Data audit log is a premium feature. You can check the StartSharp for usage sample. You don't need to use inheritance for audit log with the latest version.

**Reply by Mar-he** [reply]
So theoretically, the migration path would be:

1. Renew the enterprise license that was bought in 2020
2. Migrate to `[DataAuditLog] public class AnalysisOrderRow : Row<AnalysisOrderRow.RowFields>, IIdRow, INameRow`
3. What do I do with the AuditLogRow class? Just delete it?

In the meantime I'll write my pm a message to look for funding in some of his trousers.

**Reply by VictorTomaili** [reply]
Probably, you can inherit AuditLogRow.Fields to your MyRow.Fields inner class. 
You can implement your own data audit log functionality, or you can renew the license. It depends on you. 
If you decide to use DataAuditLog attribute, you don't need to use AuditLogRow anymore.

**Reply by VictorTomaili** [reply]
You may need to modify your AuditLogRow by removing the constructor.
